<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [119]:
import pandas as pd
import psycopg2

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [122]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

query_3_1 = f'''select 
    count(id)
    from public.vacancies
  '''


In [123]:
df = pd.read_sql_query(query_3_1, connection)
df

C:\Temp\ipykernel_7576\3851344990.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [124]:
query_3_2 = f'''select 
    count(id)
    from public.employers
  '''

In [125]:
df = pd.read_sql_query(query_3_2, connection)
df

C:\Temp\ipykernel_7576\3588252056.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [126]:
query_3_3 = f'''select 
    count(distinct name)
    from public.areas
  '''

In [127]:
df = pd.read_sql_query(query_3_3, connection)
df

C:\Temp\ipykernel_7576\2018224393.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [128]:
query_3_4 = f'''select
    count(distinct name)
    from public.industries
    '''

In [129]:
df = pd.read_sql_query(query_3_4, connection)
df

C:\Temp\ipykernel_7576\2172479966.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0    294

***

В ходе предварительного анализа данных, мы выяснили, что в исследуемой базе данных размещено 49197 вакансий, представленных 23501 работодателями из 1362 регионов. Количество сфер деятельности составляет - 294.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [130]:
query_4_1 = f'''select
    a.name,
    count(v.id)
    from public.vacancies v
    join public.areas a on v.area_id = a.id
    group by 1
    order by 2 desc
    limit 5
    '''

In [131]:
df = pd.read_sql_query(query_4_1, connection)
df

C:\Temp\ipykernel_7576\347208320.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  count
0           Москва   5333
1  Санкт-Петербург   2851
2            Минск   2112
3      Новосибирск   2006
4           Алматы   1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [132]:
query_4_2 = f'''select
    count (id)    
    from public.vacancies
where salary_from is not NULL or salary_to is not NULL
    '''

In [133]:
df = pd.read_sql_query(query_4_2, connection)
df

C:\Temp\ipykernel_7576\2168503709.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [134]:
query_4_3 = f'''select
        round (avg (salary_from)) avg_salary_from,
        round (avg (salary_to)) avg_salary_to
    from public.vacancies
    '''

In [135]:
df = pd.read_sql_query(query_4_3, connection)
df

C:\Temp\ipykernel_7576\2222372203.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



avg_salary_from  avg_salary_to
0          71065.0       110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [136]:
query_4_4 = f'''select
        count (name),
        schedule,
        employment
    from public.vacancies
    group by 2, 3
    order by 1 desc
    offset 1
    --limit 1
    '''

In [137]:
df = pd.read_sql_query(query_4_4, connection)
df

C:\Temp\ipykernel_7576\2981602351.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count          schedule           employment
0    7802  Удаленная работа     Полная занятость
1    1593     Гибкий график     Полная занятость
2    1312  Удаленная работа  Частичная занятость
3     940    Сменный график     Полная занятость
4     569       Полный день           Стажировка
5     367    Вахтовый метод     Полная занятость
6     347       Полный день  Частичная занятость
7     312     Гибкий график  Частичная занятость
8     141       Полный день     Проектная работа
9     133  Удаленная работа     Проектная работа
10    116     Гибкий график           Стажировка
11    101    Сменный график  Частичная занятость
12     64  Удаленная работа           Стажировка
13     18     Гибкий график     Проектная работа
14     12    Сменный график           Стажировка
15      2    Вахтовый метод     Проектная работа
16      1    Сменный график     Проектная работа

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [138]:
query_4_5 = f'''select
        count (id),
        experience
    from public.vacancies
    group by 2
    order by 1 
    '''

In [139]:
df = pd.read_sql_query(query_4_5, connection)
df

C:\Temp\ipykernel_7576\2601166264.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count          experience
0   1337         Более 6 лет
1   7197           Нет опыта
2  14511       От 3 до 6 лет
3  26152  От 1 года до 3 лет

***

Детальный анализ вакансий нам показал, что самым популярным регионом оказалась Москва. Что средние значения для нижней и верхней границы зарплатной вилки равны 71065 и 110537 рублей соответственно. Смогли узнать какие пары сочетаний типа рабочего графика и типа трудоустройства наиболее популярны в вакансиях. Так же узнали какой опыт работы и как часто встречается. Детальный анализ позволил выяснить, что у 24073 вакансий заполнено хотя бы одно из двух полей с зарплатой.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [140]:
query_5_1 = f'''select
        e.name,
        count(v.id)
    from public.employers e
        join public.vacancies v on e.id = v.employer_id
    group by e.id
    order by 2 desc
    limit 5    
    '''

In [141]:
df = pd.read_sql_query(query_5_1, connection)
df

C:\Temp\ipykernel_7576\1792814825.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  count
0         Яндекс   1933
1     Ростелеком    491
2       Тинькофф    444
3           СБЕР    428
4  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [142]:
query_5_2 = f'''select
        a.name,
        count(e.id) count_e,
        count(v.id) count_v
    from public.areas a
        left join public.vacancies v on a.id = v.area_id 
        left join public.employers e on e.area = a.id   
    where v.id is null        
    group by a.name
    order by count_e desc
    LIMIT 1
    '''

In [143]:
df = pd.read_sql_query(query_5_2, connection)
df

C:\Temp\ipykernel_7576\3341481329.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  count_e  count_v
0  Россия      410        0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [144]:
query_5_3 = f'''select
        e.id,
        count(distinct area_id) count_area
    from public.employers e
        left join public.vacancies v on e.id = v.employer_id 
    where v.id is not null        
    group by e.id
    order by 2 desc
    '''

In [145]:
df = pd.read_sql_query(query_5_3, connection)
df

C:\Temp\ipykernel_7576\2398953733.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



id  count_area
0         1740         181
1         2748         152
2      5724811         116
3      5130287          88
4      3682876          71
...        ...         ...
14901   810278           1
14902   810313           1
14903   810551           1
14904   810688           1
14905   811902           1

[14906 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [146]:
query_5_4 = f'''select
        count(e.id) count_employer
    from public.employers e
    left join public.employers_industries ei on e.id = ei.employer_id
    where ei.industry_id is null        
    '''

In [147]:
df = pd.read_sql_query(query_5_4, connection)
df

C:\Temp\ipykernel_7576\3114411492.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count_employer
0            8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [148]:
query_5_5 = f'''select
        e.name
    from public.employers e
    left join public.employers_industries ei on e.id = ei.employer_id
    group by 1
   having count(ei.industry_id) = 4 
   order by 1
   offset 2
   limit 1        
    '''

In [149]:
df = pd.read_sql_query(query_5_5, connection)
df

C:\Temp\ipykernel_7576\1650419958.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name
0  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [150]:
query_5_6 = f'''select
        count(ei.employer_id)
    from public.employers_industries ei
    left join public.industries i on ei.industry_id = i.id
    where i.name = 'Разработка программного обеспечения'
    '''

In [151]:
df = pd.read_sql_query(query_5_6, connection)
df

C:\Temp\ipykernel_7576\1169039133.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [152]:
import requests # код для получения списка городов-милионников
url = 'https://ru.wikipedia.org/wiki/Города-миллионеры_России'
response = requests.get(url)
from bs4 import BeautifulSoup
page = BeautifulSoup(response.text, 'html.parser') # Создаём объект BeautifulSoup, указывая html-парсер
print(page.table) # Получаем тег title, отображающийся на вкладке браузера
links = page.find_all('td')
print([link.text for link in links[3:79:5]]) #это всё до чего я смогла додуматься, хотелось бы узнать правильный способ

<table class="mbox-Обновить plainlinks metadata ambox ambox-content" role="presentation"><tbody><tr><td class="mbox-image"><div style="width:52px"><img alt="" data-file-height="620" data-file-width="620" decoding="async" height="40" src="//upload.wikimedia.org/wikipedia/commons/thumb/8/8f/Ambox_outdated_serious.svg/40px-Ambox_outdated_serious.svg.png" srcset="//upload.wikimedia.org/wikipedia/commons/thumb/8/8f/Ambox_outdated_serious.svg/60px-Ambox_outdated_serious.svg.png 1.5x, //upload.wikimedia.org/wikipedia/commons/thumb/8/8f/Ambox_outdated_serious.svg/80px-Ambox_outdated_serious.svg.png 2x" width="40"/></div></td><td class="mbox-text"><div class="mbox-text-div">Информация в этой статье или некоторых её разделах <b>устарела</b>.</div><div class="mbox-textsmall-div hide-when-compact" style="font-size:85%"><span class="hide-when-compact"> Вы можете помочь проекту, <a class="external text" href="https://ru.wikipedia.org/w/index.php?title=%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B

In [153]:
query_5_7 = f'''select
        a.name AS name,
        count(v.id) cnt 
    from public.vacancies v
        left join public.areas a ON a.id = v.area_id 
        left join public.employers e ON v.employer_id = e.id
    WHERE a.name IN ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск', 'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград')   
        AND e.name = 'Яндекс'
    GROUP BY a.id
    UNION 
    SELECT
        'total',
        COUNT(v.id)
    from public.vacancies v
        left join public.areas a ON a.id = v.area_id 
        left join public.employers e ON v.employer_id = e.id
    WHERE a.name IN ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск', 'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград') 
        AND e.name = 'Яндекс'
    ORDER BY cnt  
    '''

In [154]:
df = pd.read_sql_query(query_5_7, connection)
df

C:\Temp\ipykernel_7576\3129374555.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  cnt
0              Омск   21
1         Челябинск   22
2        Красноярск   23
3         Волгоград   24
4             Пермь   25
5            Казань   25
6    Ростов-на-Дону   25
7               Уфа   26
8            Самара   26
9         Краснодар   30
10          Воронеж   32
11      Новосибирск   35
12  Нижний Новгород   36
13     Екатеринбург   39
14  Санкт-Петербург   42
15           Москва   54
16            total  485

***

Больше всего вакансий выставляет работодатель "Яндекс", т.к. эта крупная компания имеет много различных направлений деятельности и помимо этого распрастранена по всем регионам России. Среди регионов, в которых нет вакансий, наибольшее количество работодателей. На пике популярности - удаленная работа с полной занятостью.
Проводя анализ работодателей мы можем узнать количество вакансий, выставляемых компаниями в целом, в отдельных регионах, в определенных сферах деятельности. Востребованность той или иной вакансии, работодателя, сферы деятельности и множество другой информации.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [155]:
query_6_1 = f'''select
        count(id)
    from public.vacancies
    where lower(name) LIKE '%data%'
        or lower(name) LIKE '%данн%'
    '''

In [156]:
df = pd.read_sql_query(query_6_1, connection)
df

C:\Temp\ipykernel_7576\485020324.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [157]:
query_6_2 = f'''select
        count(id) cnt
    from public.vacancies
    WHERE (name ILIKE '%data scientist%'
        OR name ILIKE '%data science%'
        OR name ILIKE '%исследователь данных%'
        OR (name ILIKE '%ML%' AND name NOT ILIKE '%HTML%') 
        OR name ILIKE '%machine learning%'
        OR name ILIKE '%машинн%обучен%')
        AND (name ILIKE '%junior%'
        OR experience = 'Нет опыта'
        OR employment = 'Стажировка')
    '''

In [158]:
df = pd.read_sql_query(query_6_2, connection)
df

C:\Temp\ipykernel_7576\3641650667.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



cnt
0   51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [159]:
query_6_3 = f'''select
        count(*) cnt
    from public.vacancies
    WHERE (name ILIKE '%data scientist%'
        OR name ILIKE '%data science%'
        OR name ILIKE '%исследователь данных%'
        OR (name LIKE '%ML%' AND name NOT ILIKE '%HTML%') 
        OR name ILIKE '%machine learning%'
        OR name ILIKE '%машинн%обучен%')
        AND (key_skills ILIKE '%SQL%'
        OR key_skills ILIKE '%postgres%')
    '''

In [160]:
df = pd.read_sql_query(query_6_3, connection)
df

C:\Temp\ipykernel_7576\1159587760.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



cnt
0  201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [161]:
query_6_4 = f'''select
        count(*) cnt
    from public.vacancies
    WHERE (name ILIKE '%data scientist%'
        OR name ILIKE '%data science%'
        OR name ILIKE '%исследователь данных%'
        OR (name LIKE '%ML%' AND name NOT ILIKE '%HTML%') 
        OR name ILIKE '%machine learning%'
        OR name ILIKE '%машинн%обучен%')
        AND key_skills ILIKE '%Python%'
    '''

In [162]:
df = pd.read_sql_query(query_6_4, connection)
df

C:\Temp\ipykernel_7576\972386973.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



cnt
0  351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [163]:
query_6_5 = f'''select
       ROUND(AVG(LENGTH(key_skills)-LENGTH(REPLACE(key_skills, CHR(9), ''))+1), 2) 
    FROM public.vacancies
    WHERE name ILIKE '%data scientist%'
        OR name ILIKE '%data science%'
        OR name ILIKE '%исследователь данных%'
        OR (name LIKE '%ML%' AND name NOT ILIKE '%HTML%') 
        OR name ILIKE '%machine learning%'
        OR name ILIKE '%машинн%обучен%'
    '''    

In [164]:
df = pd.read_sql_query(query_6_5, connection)
df

C:\Temp\ipykernel_7576\390248827.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [165]:
query_6_6 = f'''select
    ROUND(AVG(COALESCE((salary_from +salary_to)/2, salary_from, salary_to)), 0) AS mean_salary   
    from public.vacancies
    WHERE (name ILIKE '%data scientist%'
        OR name ILIKE '%data science%'
        OR name ILIKE '%исследователь данных%'
        OR (name LIKE '%ML%' AND name NOT ILIKE '%HTML%') 
        OR name ILIKE '%machine learning%'
        OR name ILIKE '%машинн%обучен%')
        AND (salary_from is not NULL or salary_to is not NULL)
        AND (experience = 'От 3 до 6 лет')
    '''    

In [166]:
df = pd.read_sql_query(query_6_6, connection)
df


C:\Temp\ipykernel_7576\2835978903.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



mean_salary
0     243115.0

***

Предметный анализ показал, что из всех вакансий исследуемой базы данных, количество вакансий, которые имеют отношение к данным - 1771, что составляет 3,6 % от всех вакансий. А количество вакансий без опыта всего лишь - 51, что говорит о высокой конкуренции и высоком пороге вхождения в данное направление(что так же подтверждает среднее количество ключевых навыков и уровень зарплаты). Так же выяснили, что знание Python более востребовано, чем SQL. 

# Общий вывод по проекту

Среди общего количества вакансий большинство вакансий сосредоточено в столицах и крупных городах. на фоне всего разнообразия сфер деятельности (а их 294), средние нижнее и верхнее значение зарплат составляют примерно 71 000 рублей и 110000 рублей. Наиболее востребованы на рынке труда сотрудники на удаленке с полной занятостью. Наиболее популярный требуемый опыт работы - щт 3 до 6 лет.

In [167]:
query_7 = f'''select
        experience,
        round (avg (salary_from)) avg_salary_from,
        round (avg (salary_to)) avg_salary_to
    from public.vacancies
    WHERE name ILIKE '%data scientist%'
        OR name ILIKE '%data science%'
        OR name ILIKE '%исследователь данных%'
        OR (name LIKE '%ML%' AND name NOT ILIKE '%HTML%') 
        OR name ILIKE '%machine learning%'
        OR name ILIKE '%машинн%обучен%'
    GROUP BY experience    
    '''
    
df = pd.read_sql_query(query_7, connection)
df

C:\Temp\ipykernel_7576\3969606264.py:15: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



experience  avg_salary_from  avg_salary_to
0         Более 6 лет              NaN            NaN
1           Нет опыта          61000.0        97000.0
2  От 1 года до 3 лет         121272.0       175673.0
3       От 3 до 6 лет         198949.0       290589.0

In [168]:
query_7_1 = f'''select
        experience,
        ROUND(AVG(COALESCE((salary_from +salary_to)/2, salary_from, salary_to)), 0) AS mean_salary   
    from public.vacancies
    WHERE (name ILIKE '%data scientist%'
        OR name ILIKE '%data science%'
        OR name ILIKE '%исследователь данных%'
        OR (name LIKE '%ML%' AND name NOT ILIKE '%HTML%') 
        OR name ILIKE '%machine learning%'
        OR name ILIKE '%машинн%обучен%')
        AND (salary_from is not NULL or salary_to is not NULL)
    GROUP BY 1       
    '''    
df_7 = pd.read_sql_query(query_7_1, connection)
df_7    

C:\Temp\ipykernel_7576\595436412.py:14: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



experience  mean_salary
0           Нет опыта      74643.0
1  От 1 года до 3 лет     139675.0
2       От 3 до 6 лет     243115.0

In [169]:
import plotly.express as px
fig = px.bar(
    data_frame=df_7,
    y='experience',
    x='mean_salary',   
    barmode='group',
    color='experience',
    orientation='h',
    title='Зависимость медианной заработной платы от готовности к переезду/командировкам'
 )
fig.show()
fig.write_html("plotly/bar_1.html")


In [170]:
query_7_2 = f'''select
        name,
        salary_from,
        salary_to
    from public.vacancies
    WHERE (name ILIKE '%data scientist%'
        OR name ILIKE '%data science%'
        OR name ILIKE '%исследователь данных%'
        OR (name LIKE '%ML%' AND name NOT ILIKE '%HTML%') 
        OR name ILIKE '%machine learning%'
        OR name ILIKE '%машинн%обучен%')
        AND (experience = 'Более 6 лет')   
    '''
    
df = pd.read_sql_query(query_7_2, connection)
df    


C:\Temp\ipykernel_7576\1122152508.py:15: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name salary_from salary_to
0                               Senior Data Scientist        None      None
1                               Senior Data Scientist        None      None
2                               Senior Data Scientist        None      None
3                               Senior Data Scientist        None      None
4                               Senior Data Scientist        None      None
5                               Senior Data Scientist        None      None
6                               Senior Data Scientist        None      None
7                               Senior Data Scientist        None      None
8                               Senior Data Scientist        None      None
9                               Senior Data Scientist        None      None
10                              Senior Data Scientist        None      None
11                              Senior Data Scientist        None      None
12                              Senior Data Scientist        None      None
13                            Data Scientist (Python)        None      None
14              Архитектор решений с опытом в Tiny ML        None      None
15              Архитектор решений с опытом в Tiny ML        None      None
16              Архитектор решений с опытом в Tiny ML        None      None
17              Архитектор решений с опытом в Tiny ML        None      None
18              Архитектор решений с опытом в Tiny ML        None      None
19              Архитектор решений с опытом в Tiny ML        None      None
20                                  Senior ML-инженер        None      None
21                                  Senior ML-инженер        None      None
22                     Machine learning research lead        None      None
23                                       ML developer        None      None
24                             Team Lead Data Science        None      None
25                                Lead Data Scientist        None      None
26                              Senior Data Scientist        None      None
27              Архитектор решений с опытов в Tiny ML        None      None
28  Machine Learning Team Lead (Go/Scala/Python) (...        None      None
29             Архитектор решений (с опытом в TinyML)        None      None
30             Архитектор решений (с опытом в TinyML)        None      None
31             Архитектор решений (с опытом в TinyML)        None      None
32             Архитектор решений (с опытом в TinyML)        None      None
33                         Machine Learning Team Lead        None      None

In [ ]:
connection.close()


Исходя из последних запросов и графика можно сделать вывод, что средняя заработная плата специалиста, работающего с данными начинается от 61 тысячи рублей и с увеличение опыта она растет практически прямо пропорцианально. А вот заработную плату с опытом работы более 6 лет компании предпочитают не указывать и этот вопрос уже решается непосредственно с соискателем.

Анализ данных показывает, что специалисты работы с данными в основном востребованы на рынке труда с хотябы минимальным опытом и с высоким уровнем требований. 